features/294-login-without-specifying-login-id-key/api.md

```diff
@@ -0,0 +1,370 @@
+login('example@example', 'password'); ==> OK
+login({
+  username: 'example',
+  email: 'example@example.com'
+}, 'password') ==> won't support this.
+```
+
+Since `loginID` is defined with `UNIQUE` constraint, which means another user **can't** signup as
+
+```
+signup({
+  username: 'example@example.com',
+  email: 'example'
+}, 'password') ==> error: Duplicated loginID
+```
+
+For app wants to use compound keys as `loginID`, we suggest it encodes compound keys to a JSON string.
+
+```
+LOGIN_SOURCES = ["email", "role_phone"]
+
+signup({
+  email: 'example@example.com',
+  role_phone: '["admin","phone"]'
+}, 'password') ==> OK
+```
+
+which creates two `lingID`s:
+
+| login_id_source:String | login_id:String (`UNIQUE`) |
+| --- | --- |
+| email | example@example.com |
+| role_phone | ["admin","phone"] |
+
+It's developer's responsibility to ensure the encoded string uniqueness ('["admin","phone"]' and '["phone","admin"]' are two different `loginID`). 
+
+And a user can't signup a `login_id_source` not in the `LOGIN_SOURCES` as well.
+
+```
+LOGIN_SOURCES = ["email", "username"]
+
+signup({
+  email: 'example@example.com',
+  role_phone: '[\"admin\", \"1234567\"]'
+}, 'password') ==> error: Unknow loginIDSource
+```
+
+The response user object is:
+
+```
+LoginID-Source: username,
+Passcode-ID: null,
+{
+    user_id: <id>,
+    login_ids: {
+      "role-phone": {
+        role: 'example',
+        phone: '8912345',
+      },
+      username: 'example',
+      email: 'example@example.com'
+    },
+    metadata: {
+      age: 18
+    }
+}
+```
+
+### Code change
+
+1. [skygear-server] Update `loginIDMetadataKeys` definition setting.
+2. [skygear-server] Change `_auth_provider_password.authData` to `text` type and add `_auth_provider_password.source` field.
+3. [skygear-JS-SDK] Update SDK APIs
+
+   | Old | New |
+   | -------- | -------- |
+   | `login(authData: dict, password: String)` | `login(loginID: String, password: String)` |
+   | N/A | `createLoginID(loginIDSource: String, loginID: String)`<br/>`updateLoginID(loginIDSource: String, loginID: String)`<br/>`removeLoginID(loginIDSource)` |
+   | N/A | `createPasscode(loginIDSource: String, password: String, skip2FA: Boolean)`<br/>`updatePasscode(loginIDSource: String, password: String, skip2FA: Boolean`<br/>`deletePasscode(loginIDSource: String)` |
+4. [skygear-server] Update signup and login handler to handle the updated `loginID` concept.
+5. [skygear-server] Include `LoginID-Source` in response header.
+
+## Compare proposal 2 and proposal 3
+
+### form of loginIDs
+
+Proposal 2 allows to use free form `loginID`s, proposal 3 not.
+
+```javascript
+signup({
+  username: 'exmaple',
+  email: 'example@example.com',
+  "role-phone": {
+    role: 'example',
+    phone: '8912345',
+  }
+}, 'password'); ===> proposal 2: OK, proposal 3: error: Incorrect loginID format
+```
+
+### Compound Keys
+
+Proposal 3 allows String loginID only, for compound keys loginID, it needs to use a work-around.
```

Comment:
Please create a feature issue to support compound LoginIDs after this proposal is merged.

Reply:
May I understand more about the feature issue? In this spec, we propose developer uses JSON string to store compound keys, in the new feature issue created after this proposal merged, does it expect we may want another mechanism to support compound keys?

Reply:
I think the features should treat compound LoginIDs as first class citizien, maybe we support JSON or other format at the API endpoint, and have the appropriate interface on each client SDK.

The end goal should be, people should feel comfortable to use LoginIDs to store the data, so they don't need to "sync" the data in loginIDs and other places (such as metadata)

And also maybe if we consider this feature seriously, we want to handle the case of "login by Role + Username or Role + Email" too.

So we drop this feature at this version, but write down the notes for everything we learnt in this discussion for the new feature, so we don't need to go through it again in future :)